Add support for agent events to Hubble API

tklauser opened the pull request:
This PR adds support for agent events to the Hubble API (`api/v1/flow.AgentEvent` and `agent_event` field in `api/v1/observer.GetFlowsRespone`). These are translated from the respective monitor API agent events (`pkg/monitor/api/AgentNotifyMessage`).

Example (ipcache upsert):
    
    $ grpcurl -plaintext -d '{"whitelist":{"event_type":{"type": 130}}}' localhost:4245 observer.Observer.GetFlows
    {
      "agentEvents": {
        "type": "IPCACHE_UPSERTED",
        "ipcacheUpdate": {
          "cidr": "10.244.1.175/32",
          "identity": 55137,
          "oldIdentity": 3,
          "hostIp": "172.18.0.2",
          "oldHostIp": "172.18.0.2",
          "namespace": "kube-system",
          "podName": "hubble-relay-5d86f9dc74-69dvt"
        }
      },
      "nodeName": "kind-control-plane",
      "time": "2020-11-24T15:02:43.535402978Z"
    }
    {
      "agentEvents": {
        "type": "IPCACHE_UPSERTED",
        "ipcacheUpdate": {
          "cidr": "172.18.0.3/32",
          "identity": 1,
          "oldIdentity": 1
        }
      },
      "nodeName": "kind-control-plane",
      "time": "2020-11-24T15:03:26.202884623Z"
    }
    [...]

See individual commit messages for more details.

Note for reviewers: I split the addition of the API types and their usage in `pkg/hubble/{parser,observer}` into multiple commits to hopefully ease review. Please let me know if you'd like me to squash these into fewer commits or put them in different order.

tklauser committed:
api/v1/flow: add AgentEvent message

Add an AgentEvent message which is used to send agent events to Hubble
clients. The agent event types match the ones currently supported by
cilium monitor (pkg/monitor/api.MessageTypeAgent), e.g.
AgentNotifyIPCacheUpserted maps to pb.AgentEventType_IPCACHE_UPSERTED.

Note that pkg/monitor/api.AgentNotifyGeneric is currently not used (and
in fact was never used) in the Cilium code base and thus not supported
by the AgentEvent message.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
api/v1/observer: add agent_event field to GetFlowsResponse message

This is used to send agent events to Hubble clients and uses the
flow.AgentEvent message introduced in the previous commit.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
hubble/parser: parse monitor API agent events into hubble API messages

This commit adds support to translate monitor API agents events as
described by pkg/monitor/api/AgentNotifyMessage into Hubble API
api/v1/flow.AgentEvent messages consumable by Hubble clients. The
monitorAPI.AgentNotify* types are translated into their corresponding
flowpb.AgentEventType_* type.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
hubble/observer: add support for flowpb.AgentEvent type

Example (ipcache upsert):

$ grpcurl -plaintext -d '{"whitelist":{"event_type":{"type": 130}}}' localhost:4245 observer.Observer.GetFlows
{
  "agentEvents": {
    "type": "IPCACHE_UPSERTED",
    "ipcacheUpdate": {
      "cidr": "10.244.1.175/32",
      "identity": 55137,
      "oldIdentity": 3,
      "hostIp": "172.18.0.2",
      "oldHostIp": "172.18.0.2",
      "namespace": "kube-system",
      "podName": "hubble-relay-5d86f9dc74-69dvt"
    }
  },
  "nodeName": "kind-control-plane",
  "time": "2020-11-24T15:02:43.535402978Z"
}
{
  "agentEvents": {
    "type": "IPCACHE_UPSERTED",
    "ipcacheUpdate": {
      "cidr": "172.18.0.3/32",
      "identity": 1,
      "oldIdentity": 1
    }
  },
  "nodeName": "kind-control-plane",
  "time": "2020-11-24T15:03:26.202884623Z"
}
[...]

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

rolinh commented on api/v1/flow/flow.proto:
```diff
@@ -472,3 +472,91 @@ message LostEvent {
+// AgentEventType is the type of agent event. These values are shared with type
+// AgentNotification in pkg/monitor/api/types.go
+enum AgentEventType {
+    UNSPECIFIED = 0;
```
Is `UNSPECIFIED == UNKNOWN` ? For other enums, we usually use the enum name+unknown.
E.g.:
```
enum Verdict {
    VERDICT_UNKNOWN = 0;
...
enum DropReason {
    // non-drop reasons
    DROP_REASON_UNKNOWN = 0;
...
enum TrafficDirection {
    TRAFFIC_DIRECTION_UNKNOWN = 0;
...
```

tklauser replied:
I took the name `UNSPECIFED` to correspond to `pkg/monitor/api.AgentNotifyUnspec`, but as far as I can tell it has the same meaning as `UNKNOWN` for other enums. Will change to `AGENT_EVENT_UNKNOWN` or `AGENT_EVENT_UNKNOWN` - both variants `*_UNKNOWN` and `UNKNOWN_*` seem to be in use. Which one is preferred for new code?

tklauser replied:
Settling for `AGENT_EVENT_UNKNOWN` as per @gandro's comment https://github.com/cilium/cilium/pull/14168#discussion_r530365049

gandro commented on api/v1/flow/flow.proto:
```diff
@@ -472,3 +472,91 @@ message LostEvent {
+// AgentEventType is the type of agent event. These values are shared with type
+// AgentNotification in pkg/monitor/api/types.go
+enum AgentEventType {
+    UNSPECIFIED = 0;
```
I know this has been copied over verbatim from pkg/monitor/api/types.go, but because protobuf enum variants are unfortunately not namespaced, please use `AGENT_EVENT_UNKNOWN` (or similar) here. 

We have historically not been consistent with the default enum variant, but I've been trying to establish the `<TYPE>_UNKNOWN` naming for newly added enums.

tklauser replied:
That answers my questions from https://github.com/cilium/cilium/pull/14168#discussion_r530385060, thanks. Will adjust to `AGENT_EVENT_UNKNOWN`.

gandro commented on api/v1/flow/flow.proto:
```diff
@@ -472,3 +472,91 @@ message LostEvent {
+// AgentNotification in pkg/monitor/api/types.go
+enum AgentEventType {
+    UNSPECIFIED = 0;
+    START = 1;
```
nit: maybe call this one `AGENT_STARTED`

Again, I know this is copied over from `pkg/monitor/api/types.go`, but in protobuf we don't have enum namespacing and so no other enum can ever call a variant `START` if we do this.

tklauser replied:
Agree, changing to `AGENT_STARTED`.

rolinh commented on pkg/hubble/parser/parser.go:
```diff
@@ -26,6 +28,8 @@ import (
 	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
 	"github.com/cilium/cilium/pkg/proxy/accesslog"
 	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/protobuf/types/known/timestamppb"
```
nit: Throughout Hubble's codebase, we use `github.com/golang/protobuf/ptypes/timestamp` which is an alias.
It would be nice to keep this consistent.

tklauser replied:
Thanks, will adjust.

gandro commented on pkg/hubble/parser/parser.go:
```diff
@@ -26,6 +28,8 @@ import (
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
 
 	"github.com/golang/protobuf/ptypes/wrappers"
```
See https://github.com/cilium/cilium/pull/14168#discussion_r530375980, this one here imports the same types as `"google.golang.org/protobuf/types/known/wrapperspb"`

tklauser replied:
Thanks. Will drop `google.golang.org/protobuf/types/known/wrapperspb`.

rolinh reviewed (commented):
This looks good to me from a first pass :+1: 

> Please let me know if you'd like me to squash these into fewer commits or put them in different order.

It's fine like this. I think your PR includes commits from the PRs referenced in the description. Note that it's possible to base your branch upon your other PR's branch. Github is smart enough to auto-switch to master once the other PR is merged (learned that from @gandro).

gandro commented on pkg/hubble/parser/parser.go:
```diff
@@ -26,6 +28,8 @@ import (
 	"github.com/cilium/cilium/pkg/proxy/accesslog"
 	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
```
Most Hubble code predates the v2 `google.golang.org/protobuf` package (c.f. https://blog.golang.org/protobuf-apiv2) and uses v1 imports instead:
```
"github.com/golang/protobuf/ptypes/timestamp"
"github.com/golang/protobuf/ptypes/wrappers"
```

Both import paths import the same types (the v1 imports are aliases for the v2 imports). We should ideally update the code base to use the new paths everywhere, but that's unrelated to the PR. 

What is important to me however is that we are consistent within a single package, so we should either remove the import of `"github.com/golang/protobuf/ptypes/wrappers"` below (see comment) or import `"github.com/golang/protobuf/ptypes/timestamp"` here.

tklauser replied:
Thanks for the explanation. Being a protobuf newbie I didn't pay much attention to this and just used the packages `goimports` added for me :see_no_evil:

Given that `github.com/golang/protobuf/ptypes/wrappers` is already imported, I'll use the v1 imports for this PR in order to stay consistent. I could send a follow-up PR to update the entire code base if you think that's something worth doing.

gandro commented on pkg/hubble/parser/parser.go:
```diff
@@ -117,6 +337,13 @@ func (p *Parser) Decode(monitorEvent *observerTypes.MonitorEvent) (*v1.Event, er
+			if !ok {
+				return nil, errors.ErrInvalidAgentMessageType
+			}
+			ev.Event = agentNotifyMessageToProto(agentNotifyMessage)
```
I don't feel strongly about this, but I think we could move this into its own package, i.e. `pkg/hubble/parser/agent`.

tklauser replied:
I was kind of the edge on whether to move this into its own package. Will move to `pkg/hubble/parser/agent`.

gandro commented on pkg/hubble/parser/parser.go:
```diff
@@ -75,6 +79,222 @@ func lostEventSourceToProto(source int) pb.LostEventSource {
+		return agentNotifyServiceDeletedToProto(pb.AgentEventType_SERVICE_DELETED, msg)
+	default:
+		return &pb.AgentEvent{
+			Type: pb.AgentEventType_UNSPECIFIED,
```
Not sure how much value this event provides as it. It just tells you that there was a unknown event, but no details or log statement as to what that event was.

I think we should either not forward unknown events at all, or at least provide some more information, e.g. `msg.Type` and maybe even the JSON encoded representation of `msg.Notification`?

tklauser replied:
Agree, will add the suggested additional information.

gandro commented on pkg/hubble/parser/parser_test.go:
```diff
@@ -28,7 +30,9 @@ import (
 	"github.com/cilium/cilium/pkg/monitor/api"
+	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
 	"github.com/cilium/cilium/pkg/proxy/accesslog"
+	"google.golang.org/protobuf/types/known/wrapperspb"
```
Aliased import of `"github.com/golang/protobuf/ptypes/wrappers"` (see above)

gandro reviewed (changes requested):
Awesome work! :rocket: 

I did a first pass and have some feedback regarding common protobuf pitfalls that I think should be addressed. Otherwise this looks very solid to me!

michi-covalent reviewed (approved):
💯

rolinh commented on api/v1/flow/flow.proto:
```diff
@@ -472,3 +472,97 @@ message LostEvent {
+    POLICY_UPDATED = 2;
+    POLICY_DELETED = 3;
+    ENDPOINT_REGENERATE_SUCCESS = 4;
+    ENDPOINT_REGENERATE_FAIL = 5;
```
nit: I'm wondering if this should be `ENDPOINT_REGENERATE_FAILURE` instead (for failure as opposed to success).

tklauser replied:
Since I somehow messed up the rebase on #14126 (CI tests are currently failing because rebasing the PR on `master` fails), I'll probably do another iteration, so will address this before that.

ti-mo reviewed (approved):
Looks great! TIL about the automagical GitHub branch retargeting, will definitely come in handy!

kaworu commented on pkg/hubble/parser/agent/parser.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+func NotifyMessageToProto(msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+	switch msg.Type {
```
Question: have you considered a type switch here on `msg.Notification` instead?

Since `msg` is only used to get its `Notification` member in the `*NotificationToProto` helper functions, we could pass the notification directly. With a type switch, we could avoid casting from all the helper functions:

```go
switch n := msg.Notification.(type) {
case monitorAPI.TimeNotification:
	if msg.Type == monitorAPI.AgentNotifyStart {
		return notifyTimeNotificationToProto(flowpb.AgentEventType_AGENT_STARTED, n)
	}
case monitorAPI.PolicyUpdateNotification:
	...
}
return notifyUnknownToProto(flowpb.AgentEventType_AGENT_EVENT_UNKNOWN, msg)
```
Would allow for

```go
func notifyTimeNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.TimeNotification) *flowpb.AgentEvent`
```

Granted the switch will be uglier, but the `*NotificationToProto` may become obviously correct and feel "safer".

tklauser replied:
Yeah, good point. I've considered this but went for the current implementation since for most `msg.Notification` members we still need to look at `msg.Type` (e.g. `monitorAPI.PolicyUpdateNotification` could be either be `monitorAPI.AgentNotifyPolicyUpdated` or `monitorAPI.AgentNotifyPolicyDeleted`) so went for a "cleaner" `NotifyMessageToProto`. But yeah, `*NotificationToProto` will be cleaner without the type switches. Given that I need to address https://github.com/cilium/cilium/pull/14168#issuecomment-734386525 anyway, I can give your proposal a go and compare again.

tklauser replied:
Changed to use a type `switch`. The proposed solution indeed feels cleaner and safer. Thanks @kAworu!


<details>
  <summary>Incremental diff:</summary>

  ```diff
diff --git a/pkg/hubble/parser/agent/parser.go b/pkg/hubble/parser/agent/parser.go
index b5a1c1488f42..c5cd4f082a05 100644
--- a/pkg/hubble/parser/agent/parser.go
+++ b/pkg/hubble/parser/agent/parser.go
@@ -27,12 +27,10 @@ import (
 	"github.com/golang/protobuf/ptypes/wrappers"
 )
 
-func notifyTimeNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+func notifyTimeNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.TimeNotification) *flowpb.AgentEvent {
 	var ts *timestamp.Timestamp
-	if notification, ok := msg.Notification.(monitorAPI.TimeNotification); ok {
-		if goTime, err := time.Parse(time.RFC3339Nano, notification.Time); err == nil {
-			ts, _ = ptypes.TimestampProto(goTime)
-		}
+	if goTime, err := time.Parse(time.RFC3339Nano, n.Time); err == nil {
+		ts, _ = ptypes.TimestampProto(goTime)
 	}
 	return &flowpb.AgentEvent{
 		Type: typ,
@@ -44,169 +42,116 @@ func notifyTimeNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.Age
 	}
 }
 
-func notifyPolicyNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var ruleCount int
-	var revision uint64
-	var labels []string
-	if n, ok := msg.Notification.(monitorAPI.PolicyUpdateNotification); ok {
-		ruleCount = n.RuleCount
-		labels = n.Labels
-		revision = n.Revision
-	}
-
+func notifyPolicyNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.PolicyUpdateNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_PolicyUpdate{
 			PolicyUpdate: &flowpb.PolicyUpdateNotification{
-				RuleCount: int64(ruleCount),
-				Labels:    labels,
-				Revision:  revision,
+				RuleCount: int64(n.RuleCount),
+				Labels:    n.Labels,
+				Revision:  n.Revision,
 			},
 		},
 	}
 }
 
-func notifyEndpointRegenNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint64
-	var labels []string
-	var e string
-	if n, ok := msg.Notification.(monitorAPI.EndpointRegenNotification); ok {
-		id = n.ID
-		labels = n.Labels
-		e = n.Error
-	}
-
+func notifyEndpointRegenNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.EndpointRegenNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_EndpointRegenerate{
 			EndpointRegenerate: &flowpb.EndpointRegenNotification{
-				Id:     id,
-				Labels: labels,
-				Error:  e,
+				Id:     n.ID,
+				Labels: n.Labels,
+				Error:  n.Error,
 			},
 		},
 	}
 }
 
-func notifyEndpointUpdateNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint64
-	var labels []string
-	var e, podName, namespace string
-	if n, ok := msg.Notification.(monitorAPI.EndpointNotification); ok {
-		id = n.ID
-		labels = n.Labels
-		e = n.Error
-		podName = n.PodName
-		namespace = n.Namespace
-	}
-
+func notifyEndpointUpdateNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.EndpointNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_EndpointUpdate{
 			EndpointUpdate: &flowpb.EndpointUpdateNotification{
-				Id:        id,
-				Labels:    labels,
-				Error:     e,
-				PodName:   podName,
-				Namespace: namespace,
+				Id:        n.ID,
+				Labels:    n.Labels,
+				Error:     n.Error,
+				PodName:   n.PodName,
+				Namespace: n.Namespace,
 			},
 		},
 	}
 }
-func notifyIPCacheNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var cidr string
-	var identity uint32
+func notifyIPCacheNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.IPCacheNotification) *flowpb.AgentEvent {
 	var oldIdentity *wrappers.UInt32Value
-	var hostIPString, oldHostIPString string
-	var encryptKey uint8
-	var namespace, podName string
-	if n, ok := msg.Notification.(monitorAPI.IPCacheNotification); ok {
-		cidr = n.CIDR
-		identity = n.Identity
-		if n.OldIdentity != nil {
-			oldIdentity = &wrappers.UInt32Value{
-				Value: *n.OldIdentity,
-			}
-		}
-		if n.HostIP != nil {
-			hostIPString = n.HostIP.String()
+	if n.OldIdentity != nil {
+		oldIdentity = &wrappers.UInt32Value{
+			Value: *n.OldIdentity,
 		}
-		if n.OldHostIP != nil {
-			oldHostIPString = n.OldHostIP.String()
-		}
-		encryptKey = n.EncryptKey
-		namespace = n.Namespace
-		podName = n.PodName
+	}
+	var hostIPString string
+	if n.HostIP != nil {
+		hostIPString = n.HostIP.String()
+	}
+	var oldHostIPString string
+	if n.OldHostIP != nil {
+		oldHostIPString = n.OldHostIP.String()
 	}
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_IpcacheUpdate{
 			IpcacheUpdate: &flowpb.IPCacheNotification{
-				Cidr:        cidr,
-				Identity:    identity,
+				Cidr:        n.CIDR,
+				Identity:    n.Identity,
 				OldIdentity: oldIdentity,
 				HostIp:      hostIPString,
 				OldHostIp:   oldHostIPString,
-				EncryptKey:  uint32(encryptKey),
-				Namespace:   namespace,
-				PodName:     podName,
+				EncryptKey:  uint32(n.EncryptKey),
+				Namespace:   n.Namespace,
+				PodName:     n.PodName,
 			},
 		},
 	}
 }
 
-func notifyServiceUpsertedToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint32
-	var feAddr *flowpb.ServiceUpsertNotificationAddr
-	var beAddrs []*flowpb.ServiceUpsertNotificationAddr
-	var svcType, svcTrafficPolicy, svcName, svcNamespace string
-	if n, ok := msg.Notification.(monitorAPI.ServiceUpsertNotification); ok {
-		id = n.ID
-		feAddr = &flowpb.ServiceUpsertNotificationAddr{
-			Ip:   n.Frontend.IP.String(),
-			Port: uint32(n.Frontend.Port),
-		}
-		for _, be := range n.Backends {
-			var ipStr string
-			if be.IP != nil {
-				ipStr = be.IP.String()
-			}
-			beAddrs = append(beAddrs, &flowpb.ServiceUpsertNotificationAddr{
-				Ip:   ipStr,
-				Port: uint32(be.Port),
-			})
+func notifyServiceUpsertedToProto(typ flowpb.AgentEventType, n monitorAPI.ServiceUpsertNotification) *flowpb.AgentEvent {
+	feAddr := &flowpb.ServiceUpsertNotificationAddr{
+		Ip:   n.Frontend.IP.String(),
+		Port: uint32(n.Frontend.Port),
+	}
+	beAddrs := make([]*flowpb.ServiceUpsertNotificationAddr, 0, len(n.Backends))
+	for _, be := range n.Backends {
+		var ipStr string
+		if be.IP != nil {
+			ipStr = be.IP.String()
 		}
-		svcType = n.Type
-		svcTrafficPolicy = n.TrafficPolicy
-		svcName = n.Name
-		svcNamespace = n.Namespace
+		beAddrs = append(beAddrs, &flowpb.ServiceUpsertNotificationAddr{
+			Ip:   ipStr,
+			Port: uint32(be.Port),
+		})
 	}
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_ServiceUpsert{
 			ServiceUpsert: &flowpb.ServiceUpsertNotification{
-				Id:               id,
+				Id:               n.ID,
 				FrontendAddress:  feAddr,
 				BackendAddresses: beAddrs,
-				Type:             svcType,
-				TrafficPolicy:    svcTrafficPolicy,
-				Name:             svcName,
-				Namespace:        svcNamespace,
+				Type:             n.Type,
+				TrafficPolicy:    n.TrafficPolicy,
+				Name:             n.Name,
+				Namespace:        n.Namespace,
 			},
 		},
 	}
 }
 
-func notifyServiceDeletedToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint32
-	if n, ok := msg.Notification.(monitorAPI.ServiceDeleteNotification); ok {
-		id = n.ID
-	}
+func notifyServiceDeletedToProto(typ flowpb.AgentEventType, n monitorAPI.ServiceDeleteNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_ServiceDelete{
 			ServiceDelete: &flowpb.ServiceDeleteNotification{
-				Id: id,
+				Id: n.ID,
 			},
 		},
 	}
@@ -226,30 +171,43 @@ func notifyUnknownToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyM
 }
 
 func NotifyMessageToProto(msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	switch msg.Type {
-	case monitorAPI.AgentNotifyStart:
-		return notifyTimeNotificationToProto(flowpb.AgentEventType_AGENT_STARTED, msg)
-	case monitorAPI.AgentNotifyPolicyUpdated:
-		return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_UPDATED, msg)
-	case monitorAPI.AgentNotifyPolicyDeleted:
-		return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_DELETED, msg)
-	case monitorAPI.AgentNotifyEndpointRegenerateSuccess:
-		return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, msg)
-	case monitorAPI.AgentNotifyEndpointRegenerateFail:
-		return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_FAILURE, msg)
-	case monitorAPI.AgentNotifyEndpointCreated:
-		return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_CREATED, msg)
-	case monitorAPI.AgentNotifyEndpointDeleted:
-		return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_DELETED, msg)
-	case monitorAPI.AgentNotifyIPCacheUpserted:
-		return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_UPSERTED, msg)
-	case monitorAPI.AgentNotifyIPCacheDeleted:
-		return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_DELETED, msg)
-	case monitorAPI.AgentNotifyServiceUpserted:
-		return notifyServiceUpsertedToProto(flowpb.AgentEventType_SERVICE_UPSERTED, msg)
-	case monitorAPI.AgentNotifyServiceDeleted:
-		return notifyServiceDeletedToProto(flowpb.AgentEventType_SERVICE_DELETED, msg)
-	default:
-		return notifyUnknownToProto(flowpb.AgentEventType_AGENT_EVENT_UNKNOWN, msg)
+	switch n := msg.Notification.(type) {
+	case monitorAPI.TimeNotification:
+		if msg.Type == monitorAPI.AgentNotifyStart {
+			return notifyTimeNotificationToProto(flowpb.AgentEventType_AGENT_STARTED, n)
+		}
+	case monitorAPI.PolicyUpdateNotification:
+		if msg.Type == monitorAPI.AgentNotifyPolicyUpdated {
+			return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_UPDATED, n)
+		} else if msg.Type == monitorAPI.AgentNotifyPolicyDeleted {
+			return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_DELETED, n)
+		}
+	case monitorAPI.EndpointRegenNotification:
+		if msg.Type == monitorAPI.AgentNotifyEndpointRegenerateSuccess {
+			return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, n)
+		} else if msg.Type == monitorAPI.AgentNotifyEndpointRegenerateFail {
+			return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_FAILURE, n)
+		}
+	case monitorAPI.EndpointNotification:
+		if msg.Type == monitorAPI.AgentNotifyEndpointCreated {
+			return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_CREATED, n)
+		} else if msg.Type == monitorAPI.AgentNotifyEndpointDeleted {
+			return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_DELETED, n)
+		}
+	case monitorAPI.IPCacheNotification:
+		if msg.Type == monitorAPI.AgentNotifyIPCacheUpserted {
+			return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_UPSERTED, n)
+		} else if msg.Type == monitorAPI.AgentNotifyIPCacheDeleted {
+			return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_DELETED, n)
+		}
+	case monitorAPI.ServiceUpsertNotification:
+		if msg.Type == monitorAPI.AgentNotifyServiceUpserted {
+			return notifyServiceUpsertedToProto(flowpb.AgentEventType_SERVICE_UPSERTED, n)
+		}
+	case monitorAPI.ServiceDeleteNotification:
+		if msg.Type == monitorAPI.AgentNotifyServiceDeleted {
+			return notifyServiceDeletedToProto(flowpb.AgentEventType_SERVICE_DELETED, n)
+		}
 	}
+	return notifyUnknownToProto(flowpb.AgentEventType_AGENT_EVENT_UNKNOWN, msg)
 }
diff --git a/pkg/hubble/parser/agent/parser_test.go b/pkg/hubble/parser/agent/parser_test.go
index 4ad7273ad8b9..0d97c99badef 100644
--- a/pkg/hubble/parser/agent/parser_test.go
+++ b/pkg/hubble/parser/agent/parser_test.go
@@ -75,7 +75,7 @@ func TestDecodeAgentEvent(t *testing.T) {
 		ev   *flowpb.AgentEvent
 	}{
 		{
-			name: "empty AgentNotifyMessage",
+			name: "empty",
 			msg:  api.AgentNotifyMessage{},
 			ev: &flowpb.AgentEvent{
 				Type: flowpb.AgentEventType_AGENT_EVENT_UNKNOWN,
@@ -88,7 +88,7 @@ func TestDecodeAgentEvent(t *testing.T) {
 			},
 		},
 		{
-			name: "unspecified AgentNotifyMessage",
+			name: "unspecified",
 			msg: api.AgentNotifyMessage{
 				Type:         api.AgentNotifyUnspec,
 				Notification: unknownNotification,
@@ -103,6 +103,23 @@ func TestDecodeAgentEvent(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "type and notification type mismatch",
+			msg: api.AgentNotifyMessage{
+				Type:         api.AgentNotifyStart,
+				Notification: unknownNotification,
+			},
+			ev: &flowpb.AgentEvent{
+				Type: flowpb.AgentEventType_AGENT_EVENT_UNKNOWN,
+				Notification: &flowpb.AgentEvent_Unknown{
+					Unknown: &flowpb.AgentEventUnknown{
+						Type:         fmt.Sprintf("%d", api.AgentNotifyStart),
+						Notification: string(unknownNotificationJSON),
+					},
+				},
+			},
+		},
+
 		{
 			name: "StartMessage",
 			msg:  api.StartMessage(agentStartTS),
```
</details>

kaworu commented on pkg/hubble/parser/parser_test.go:
```diff
@@ -29,7 +29,6 @@ import (
 	"github.com/cilium/cilium/pkg/monitor"
 	"github.com/cilium/cilium/pkg/monitor/api"
 	"github.com/cilium/cilium/pkg/proxy/accesslog"
-
```
Is this removal intended?

tklauser replied:
This was an unintentional change when I moved tests to `pkg/hubble/parser/agent`. Will revert, thanks.

tklauser committed:
hubble/observer: use event timestamp when a time range is set

When r.timeRange is set but the received event is not a v1.Flow (i.e.
e.GetFlow() is nil), v1.Flow(nil).GetTime() will return nil and lead to
a panic. So far, this wasn't triggered in CI because the only other
option for e.Event.(type) was *flowpb.LostEvent. These are very uncommon
and hard to trigger in CI, thus the failure was not observed to far.

Fix this by using e.Timestamp consistently which is always set.

Reference: https://github.com/cilium/cilium/pull/14168#issuecomment-734386525
Reference: https://github.com/cilium/cilium/pull/14168#issuecomment-734405743
Fixes: f40c4ce4ad65 ("hubble: Import server-side Hubble code")
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

kaworu reviewed (approved):
Thanks for this, LGTM!
